rig/machine_control/machine_controller.py

```diff
@@ -1739,34 +1807,169 @@ def get_machine(self, x=0, y=0, default_num_cores=18):
-                        num_working_cores = self.get_num_working_cores(x, y)
-                        working_links = self.get_working_links(x, y)
-
-                        if num_working_cores < default_num_cores:
-                            resource_exception = chip_resources.copy()
-                            resource_exception[Cores] = min(default_num_cores,
-                                                            num_working_cores)
-                            chip_resource_exceptions[(x, y)] = \
-                                resource_exception
-
-                        for link in set(Links) - working_links:
-                            dead_links.add((x, y, link))
-                    except SCPError:
-                        # The chip was listed in the P2P table but is not
-                        # responding. Assume it is dead anyway.
-                        dead_chips.add((x, y))
+                    if chip_info.num_cores < default_num_cores:
+                        resource_exception = chip_resources.copy()
+                        resource_exception[Cores] = min(default_num_cores,
+                                                        chip_info.num_cores)
+                        chip_resource_exceptions[(x, y)] = \
+                            resource_exception
 
-        return Machine(max_x + 1, max_y + 1,
+                    for link in set(Links) - chip_info.working_links:
+                        dead_links.add((x, y, link))
+
+        return Machine(width, height,
                        chip_resources,
                        chip_resource_exceptions,
                        dead_chips, dead_links)
 
+    @ContextMixin.use_contextual_arguments()
+    def get_resource_constraints(self, machine,
+                                 core_resource=Cores,
+                                 sdram_resource=SDRAM,
+                                 sram_resource=SRAM,
+                                 _all_chips_info=None):
+        """Returns a set of place-and-route
+        :py:class:`~rig.place_and_route.constraints` which reserves any cores
+        and SDRAM that are already in use.
+
+        .. note::
+            Most users will probably prefer to use
+            :py:meth:`.get_machine_and_resource_constraints` which both gets a
+            :py:class:`rig.machine.Machine` object and a set of place-and-route
+            constraints with reduced overhead over calling both methods
+            individually.
+
+        This method, like :py:meth:`get_machine`, is aimed at users of Rig's
+        place-and-route tools and assumes that one resource type represents
+        Cores while another represents SDRAM.
+
+        The returned set of
+        :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`s
+        reserves all cores not in an Idle state (i.e. not a monitor and not
+        already running an application).
+
+        This set also reserves ranges of SDRAM/SRAM starting at the top
```

Comment:
This set also contains reservations to SDRAM/SRAM to account for any applications already making use of the memory.?